### PR TITLE
Coco output from detectron predictions - Py script

### PR DESCRIPTION
**Pull request by sahandv**
## About:

Added COCO output to the predictor.

* This will scan a directory for a file pattern and do predictions on all of them, and then will output a single coco JSON file. 
* Using this function will not generate visualisations, and will only generate actual outputs to be plugged into [aerial-conversion/scripts/coco2geojson.py](https://github.com/Sydney-Informatics-Hub/aerial-conversion#usage).
* Shape simplification is added, which can be controlled with a tolerance argument `-s`. This reduces the output JSON size by a large scale.

For usage:

``` bash
python -m scripts.prediction_batch_detectron2 -i "path/to/tiles" -c "path/to/config.yml" -w "path/to/weights/model.pth" -s 0.3 --force-cpu 
```
Deliberately added an `__init__.py` to scripts, so we can run them as modules from the project root.

This satisfies #24, and partially satisfies [Shape regularisation issue in aerial-conversion at #19](https://github.com/Sydney-Informatics-Hub/aerial-conversion/issues/19).

---
## Note:
* This functionality does not use the batch processing built into Detectron2. It just iterates over a list of images.
* A basic testing was added. Can be further improved in future revisions.
---

## ~~WIP~~: 
- [x] unit testing
- [x] shape simplification

### CODE REVIEW

File: scripts/prediction_batch_detectron2.py
Comment (xinwei89) [inline scripts/prediction_batch_detectron2.py:1]: Is this code converted from https://github.com/Sydney-Informatics-Hub/aerial-segmentation/pull/28

File: aerialseg/utils.py [hunk 1]
```diff
@@ -26,4 +40,186 @@ def save_images_as_gif(input_folder, output_gif_path, duration=100):
...
+                polygon = polygon_prep(p, simplify_tolerance=simplify_tolerance)
+                mask_arrays.append(mask_array_instance)
+                labels_list.append(labels[i])
+                bbox_list.append(bbox[i])
+
+                if flatten:
+                    polygons.append(polygon.flatten().tolist())
+                else:
+                    polygons.append(polygon.tolist())
+
```
[the remark this answers is not in the document]
Comment (sahandv) [reply aerialseg/utils.py:134]: Good catch. Thanks

File: scripts/prediction_batch_detectron2.py
Comment (sahandv) [reply scripts/prediction_batch_detectron2.py:1]: This code is converted from #28 and `prediction_detectron2.py`.

Comment (xinwei89) [reply scripts/prediction_batch_detectron2.py:1]: Do you still need review on https://github.com/Sydney-Informatics-Hub/aerial-segmentation/pull/28 in that case

File: scripts/prediction_batch_detectron2.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        with open(args.coco, "r") as f:
+            coco = json.load(f)
+        categories = coco["categories"]
+        categories_keyed = {
+            c["id"]: {"name": c["name"], "supercategory": c["supercategory"]}
+            for c in categories
+        }
+        # thing_classes = [c["name"] for c in categories] # These 3 steps are not required for inference if not visualising
+        # meta = MetadataCatalog.get("predict")
+        # meta.thing_classes = thing_classes
```
Comment (xinwei89) [inline scripts/prediction_batch_detectron2.py:108]: I agree, `meta` is not used elsewhere. The categories can be extracted from the coco file.

File: scripts/prediction_batch_detectron2.py
Comment (sahandv) [reply scripts/prediction_batch_detectron2.py:1]: #28 is no longer required. I added a wontfix tag to #28.

File: scripts/prediction_batch_detectron2.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+# from detectron2.data import MetadataCatalog
+from detectron2.engine import DefaultPredictor
+
+from aerialseg.utils import assemble_coco_json, extract_all_annotations_df
+
+
+def create_parser():
+    parser = argparse.ArgumentParser(
+        description="Run Detectron2 prediction on aerial imagery and show image with results."
+    )
```
Comment (xinwei89) [inline scripts/prediction_batch_detectron2.py:20]: Hmmm, is this the correct description? Don't think the batch script visualizes.

File: README.md
Comment (xinwei89) [inline README.md:1]: Maybe add a visualize function in here... so we can reuse it as needed and don't need to do it separately like in `prediction_detectron2.py`

File: aerialseg/utils.py
Comment (xinwei89) [inline aerialseg/utils.py:1]: Maybe add a visualize function in `utils.py` ... so we can reuse it as needed and don't need to do it separately like in `prediction_detectron2.py`

Comment (sahandv) [reply aerialseg/utils.py:1]: Referenced in #32 .

**Review by xinwei89 — APPROVED** [review]

**Pull request merged**